### PR TITLE
Revamped the BabelFlow PMT filter to be able to compute relevance

### PR DESCRIPTION
**Pull request by sshudler**
This PR revamps the PMT filter by adding a dataflow that can compute the relevance. This computation is switched on by an additional flag. Also added a unit-test for this new functionality.

Assumes that we use BabelFlow with tag v1.1.0 and PMT with tag v1.1.0.

### CODE REVIEW

**Comment by cyrush** [comment]
@sshudler  -- hit some compiler errors when building pmt 1.1.0. Looks like missing headers, probably easy to resolve. 

```
[ 93%] Building CXX object src/CMakeFiles/pmt.dir/AugmentedMergeTree.cpp.o
cd /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-build-myumokp/src && /usr/bin/mpicxx -Dpmt_EXPORTS -I/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src -isystem /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/babelflow-1.1.0-tokcztolfbwrduipoyftao2na25p7sxs/include -std=c++11 -O2 -g -DNDEBUG -fPIC -MD -MT src/CMakeFiles/pmt.dir/MergeTree.cpp.o -MF CMakeFiles/pmt.dir/MergeTree.cpp.o.d -o CMakeFiles/pmt.dir/MergeTree.cpp.o -c /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/MergeTree.cpp
cd /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-build-myumokp/src && /usr/bin/mpicxx -Dpmt_EXPORTS -I/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src -isystem /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/babelflow-1.1.0-tokcztolfbwrduipoyftao2na25p7sxs/include -std=c++11 -O2 -g -DNDEBUG -fPIC -MD -MT src/CMakeFiles/pmt.dir/UnionFind.cpp.o -MF CMakeFiles/pmt.dir/UnionFind.cpp.o.d -o CMakeFiles/pmt.dir/UnionFind.cpp.o -c /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/UnionFind.cpp
cd /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-build-myumokp/src && /usr/bin/mpicxx -Dpmt_EXPORTS -I/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src -isystem /uberenv_libs/spack/opt/spack/linux-ubuntu18.04-zen/gcc-7.5.0/babelflow-1.1.0-tokcztolfbwrduipoyftao2na25p7sxs/include -std=c++11 -O2 -g -DNDEBUG -fPIC -MD -MT src/CMakeFiles/pmt.dir/AugmentedMergeTree.cpp.o -MF CMakeFiles/pmt.dir/AugmentedMergeTree.cpp.o.d -o CMakeFiles/pmt.dir/AugmentedMergeTree.cpp.o -c /tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/AugmentedMergeTree.cpp
/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/MergeTree.cpp: In member function 'void MergeTree::writeToFile(uint32_t)':
/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/MergeTree.cpp:637:61: warning: format '%llu' expects argument of type 'long long unsigned int', but argument 3 has type 'BabelFlow::GlobalIndexType {aka unsigned int}' [-Wformat=]
       fprintf(fp, "%llu->%llu\n", it->id(), it->down()->id());
                                   ~~~~~~~~                  ^
/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/MergeTree.cpp:637:61: warning: format '%llu' expects argument of type 'long long unsigned int', but argument 4 has type 'BabelFlow::GlobalIndexType {aka unsigned int}' [-Wformat=]
/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/SortedJoinAlgorithm.cpp: In function 'int compute_relevance_field(MergeTree&, BabelFlow::GlobalIndexType*, BabelFlow::FunctionType*, BabelFlow::FunctionType*, BabelFlow::GlobalIndexType)':
/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-src/src/SortedJoinAlgorithm.cpp:302:8: error: 'stable_sort' is not a member of 'std'
   std::stable_sort(node_ids.begin(), node_ids.end(),
        ^~~~~~~~~~~
src/CMakeFiles/pmt.dir/build.make:218: recipe for target 'src/CMakeFiles/pmt.dir/SortedJoinAlgorithm.cpp.o' failed
make[2]: *** [src/CMakeFiles/pmt.dir/SortedJoinAlgorithm.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-build-myumokp'
CMakeFiles/Makefile2:118: recipe for target 'src/CMakeFiles/pmt.dir/all' failed
make[1]: *** [src/CMakeFiles/pmt.dir/all] Error 2
make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-parallelmergetree-1.1.0-myumokplmhi5adva7tn2qnjti6xuqbyf/spack-build-myumokp'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by cyrush** [comment]
I saw spack pkg has case to help fix this for gcc 11+, however I think I saw this with gcc 10 (trying again)

File: src/ascent/runtimes/flow_filters/ascent_runtime_babelflow_pmt.cpp [hunk 1]
```diff
@@ -674,14 +1016,25 @@ void ascent::runtime::filters::BFlowPmt::execute()
     int64_t fanin = p["fanin"].as_int64();
     FunctionType threshold = p["threshold"].as_float64();
     int64_t gen_field = p["gen_segment"].as_int64();
+    int64_t gen_rel = p["rel_field"].as_int64();
```
Comment (cyrush) [inline src/ascent/runtimes/flow_filters/ascent_runtime_babelflow_pmt.cpp]: We now have updated containers, -- looks like this  line is failing during the unit test runs

```
77: Node::as_int64() const -- DataType empty at path rel_field does not equal expected DataType int64
```

full log:
[2021_12_blfow_run_log.txt](https://github.com/Alpine-DAV/ascent/files/7794394/2021_12_blfow_run_log.txt)

**Comment by sshudler** [comment]
@cyrush I'm sorry for the delay - I was away for most of February. The compilation error in pmt 1.1.0 is because I forgot to add `#include <algorithm>`. It compiled with clang on my Mac, but with gcc 10 it gives me this error. I'll fix that and create a new tag. How were you able to compile the code and run the unit tests? Is adding the fixed tag a prerequisite for the merge?

File: src/ascent/runtimes/flow_filters/ascent_runtime_babelflow_pmt.cpp (hunk 1 above)
Comment (sshudler) [reply src/ascent/runtimes/flow_filters/ascent_runtime_babelflow_pmt.cpp]: @cyrush It's a good point - the bug is that `rel_field` is an optional parameter and I don't check whether it's defined or not. I pushed a fix now.

**Comment by cyrush** [comment]
closing in favor of #945 which was merged.

**Pull request closed without merging**